boa/src/object/gcobject.rs

```diff
@@ -176,31 +206,106 @@ impl GcObject {
+                            let second_env = FunctionEnvironmentRecord::new(
+                                this_function_object,
+                                if construct || !flags.is_lexical_this_mode() {
+                                    Some(this)
+                                } else {
+                                    None
+                                },
+                                Some(local_env),
+                                // Arrow functions do not have a this binding https://tc39.es/ecma262/#sec-function-environment-records
+                                if flags.is_lexical_this_mode() {
+                                    BindingStatus::Lexical
+                                } else {
+                                    BindingStatus::Uninitialized
+                                },
+                                Value::undefined(),
+                            );
+                            context.push_environment(second_env);
+                        }
 
                         FunctionBody::Ordinary(body.clone())
                     }
                 }
-            } else {
-                return context.throw_type_error("function object is not callable");
             }
         } else {
             return context.throw_type_error("not a function");
         };
 
-        match f_body {
-            FunctionBody::BuiltInFunction(func) => func(this, args, context),
-            FunctionBody::BuiltInConstructor(func) => func(&Value::undefined(), args, context),
+        match body {
+            FunctionBody::BuiltInConstructor(function) if construct => {
+                function(&this_target, args, context)
+            }
+            FunctionBody::BuiltInFunction(_) if construct => {
+                unreachable!("Cannot have a function in construct")
+            }
+            FunctionBody::BuiltInConstructor(function) => {
+                function(&Value::undefined(), args, context)
+            }
+            FunctionBody::BuiltInFunction(function) => function(this_target, args, context),
             FunctionBody::Ordinary(body) => {
                 let result = body.run(context);
+                let this = context.get_this_binding();
+
+                if has_parameter_expressions {
+                    context.pop_environment();
+                }
                 context.pop_environment();
 
-                result
+                if construct {
+                    this
+                } else {
+                    result
+                }
             }
         }
     }
 
+    /// Call this object.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the object is currently mutably borrowed.
+    // <https://tc39.es/ecma262/#sec-prepareforordinarycall>
+    // <https://tc39.es/ecma262/#sec-ecmascript-function-objects-call-thisargument-argumentslist>
+    #[track_caller]
+    pub fn call(&self, this: &Value, args: &[Value], context: &mut Context) -> Result<Value> {
```

Comment:
This should improve performance a bit in debug builds:

```suggestion
    #[inline]
    pub fn call(&self, this: &Value, args: &[Value], context: &mut Context) -> Result<Value> {
```